src/foam/nanos/notification/email/SMTPEmailService.js

```diff
@@ -190,27 +193,45 @@ foam.CLASS({
       `
     },
     {
-      name: 'sendEmail',
+      name: 'execute',
       javaCode:
       `
-      OMLogger omLogger = (OMLogger) x.get("OMLogger");
-      Logger logger = (Logger) getX().get("logger");
-        try {
-          omLogger.log("Pre send email request");
-          MimeMessage message = createMimeMessage(emailMessage);
-          Transport transport = getSession_().getTransport("smtp");
-          transport.connect();
-          logger.info("SMTPEmailService connected.");
-          transport.send(message, getUsername(), getPassword());
-          logger.info("SMTPEmailService sent MimeMessage.");
-          transport.close();
-          logger.info("SMTPEmailService finish.");
-          omLogger.log("Post send email request");
-        } catch (Exception e) {
-          logger.error("SMTPEmailService failed to finish. " + e);
-        }
+        DAO emailMessageDAO = (DAO) x.get("emailMessageDAO");
+        List<EmailMessage> emailMessages = ((foam.dao.ArraySink)
+          emailMessageDAO
+            .where(EQ(EmailMessage.STATUS, Status.UNSENT))
+            .select(new ArraySink()))
+            .getArray();
         
-      `
+            OMLogger omLogger = (OMLogger) x.get("OMLogger");
+            Logger logger = (Logger) getX().get("logger");
+        
+            try {
+              omLogger.log("Pre send email request");
+              Transport transport = getSession_().getTransport("smtp");
+              transport.connect();
+              logger.info("SMTPEmailService connected.");
+              for (EmailMessage emailMessage : emailMessages) {
+                MimeMessage message = createMimeMessage(emailMessage);
+                try {
+                  transport.send(message, getUsername(), getPassword());
+                  emailMessage.setStatus(Status.SENT);
+                  logger.info("SMTPEmailService sent MimeMessage.");
+                } catch ( MessagingException e ) {
+                  emailMessage.setStatus(Status.FAILED);
```

Comment:
If we experience a transport failure or the connection is closed, break the for loop.